turbopack/crates/turbopack-ecmascript/src/lib.rs

```diff
@@ -974,49 +1046,762 @@ impl EcmascriptModuleContent {
+            })
+            .map(|m| m.chunk_item_id(*options.chunking_context).to_resolved())
+            .try_join()
+            .await?
+            .into();
+
+        emit_content(content, additional_ids).await
+    }
+}
+
+/// Merges multiple Ecmascript modules into a single AST, setting the syntax contexts correctly so
+/// that imports work.
+///
+/// In `contents`, each import from another module in the group must have an Ident with
+/// - a `ctxt` listed in scope_hoisting_syntax_contexts.module_contexts, and
+/// - `sym` being the name of the import.
+///
+/// This is then used to map back to the variable name and context of the exporting module.
+#[allow(clippy::type_complexity)]
+async fn merge_modules(
+    mut contents: Vec<(ResolvedVc<Box<dyn EcmascriptChunkPlaceable>>, CodeGenResult)>,
+    entry_points: &Vec<(ResolvedVc<Box<dyn EcmascriptChunkPlaceable>>, usize)>,
+    globals_merged: &'_ Globals,
+) -> Result<(
+    Program,
+    Vec<CodeGenResultComments>,
+    Vec<CodeGenResultSourceMap>,
+    SmallVec<[ResolvedVc<Box<dyn GenerateSourceMap>>; 1]>,
+)> {
+    struct SetSyntaxContextVisitor<'a> {
+        modules_width: u32,
+        current_module: ResolvedVc<Box<dyn EcmascriptChunkPlaceable>>,
+        current_module_idx: u32,
+        /// The export syntax contexts in the current AST, which will be mapped to merged_ctxts
+        reverse_module_contexts:
+            FxIndexMap<SyntaxContext, ResolvedVc<Box<dyn EcmascriptChunkPlaceable>>>,
+        export_contexts:
+            &'a FxHashMap<ResolvedVc<Box<dyn EcmascriptChunkPlaceable>>, &'a FxHashMap<RcStr, Id>>,
+        /// A fresh global SyntaxContext for each module-local context, so that we can merge them
+        /// into a single global AST.
+        unique_contexts_cache: &'a mut FxHashMap<
+            (ResolvedVc<Box<dyn EcmascriptChunkPlaceable>>, SyntaxContext),
+            SyntaxContext,
+        >,
+    }
+
+    impl<'a> SetSyntaxContextVisitor<'a> {
+        fn get_context_for(
+            &mut self,
+            module: ResolvedVc<Box<dyn EcmascriptChunkPlaceable>>,
+            local_ctxt: SyntaxContext,
+        ) -> SyntaxContext {
+            if let Some(&global_ctxt) = self.unique_contexts_cache.get(&(module, local_ctxt)) {
+                global_ctxt
+            } else {
+                let global_ctxt = SyntaxContext::empty().apply_mark(Mark::new());
+                self.unique_contexts_cache
+                    .insert((module, local_ctxt), global_ctxt);
+                global_ctxt
+            }
+        }
+    }
+
+    impl VisitMut for SetSyntaxContextVisitor<'_> {
+        fn visit_mut_ident(&mut self, ident: &mut Ident) {
+            let Ident {
+                sym, ctxt, span, ..
+            } = ident;
+
+            if let Some(&module) = self.reverse_module_contexts.get(ctxt) {
+                let eval_context_exports = self.export_contexts.get(&module).unwrap();
+                // TODO looking up an Atom in a Map<RcStr, _>
```

Comment:
Atom and RcStr have very nearly identical definitions.

I wonder if we should just endorse a `transmute` solution somewhere...